### PR TITLE
feat: monitor firebase messaging token changes and update at the server

### PR DESCRIPTION
**Pull request by alexis-kochetov**

### CODE REVIEW

File: src/dotnet/App.Maui/Platforms/Android/Notifications/FirebaseMessagingService.cs [hunk 1]
```diff
@@ -38,6 +44,17 @@ public class FirebaseMessagingService : Firebase.Messaging.FirebaseMessagingServ
     public override void OnNewToken(string token)
     {
         Log.LogDebug("OnNewToken: '{Token}'", token);
+        if (TryGetScopedServices(out var services)) {
```
Comment (crui3er) [inline src/dotnet/App.Maui/Platforms/Android/Notifications/FirebaseMessagingService.cs]: ScopedServices are initialized when MauiBlazorApp is initialized, so UI is loaded. For case when we have UI loaded, NotificationsPermissionBanner invokes registering device.
If we want to react on new token message when ui is not loaded we need to use services from root container only.
Also I think we should check if notification permission is granted.

**Review by crui3er — CHANGES_REQUESTED** [review]

File: src/dotnet/App.Maui/Platforms/Android/Notifications/FirebaseMessagingService.cs (hunk 1 above)
Comment (alexis-kochetov) [reply src/dotnet/App.Maui/Platforms/Android/Notifications/FirebaseMessagingService.cs]: We don't have a session in the root container, so do you have any ideas 🤔

File: src/dotnet/App.Maui/Platforms/iOS/AppDelegate.cs [hunk 2]
```diff
@@ -42,6 +52,28 @@ public override void DidEnterBackground(UIApplication application)
...
+            return;
+
+        var rpcHub = appServices.GetService<RpcHub>();
+        var commander = appServices.GetService<ICommander>();
+        var sessionResolver = appServices.GetRequiredService<TrueSessionResolver>();
+        if (rpcHub != null && commander != null)
+            _ = Task.Run(async () => {
+                await rpcHub.WhenClientPeerConnected().ConfigureAwait(false);
+                var session = await sessionResolver.GetSession().ConfigureAwait(false);
+                var command = new Notifications_RegisterDevice(session, token, DeviceType.AndroidApp);
```
Comment (crui3er) [inline src/dotnet/App.Maui/Platforms/iOS/AppDelegate.cs]: Use DeviceType.iOSApp

File: src/dotnet/App.Maui/Platforms/iOS/AppDelegate.cs [hunk 3]
```diff
@@ -42,6 +52,28 @@ public override void DidEnterBackground(UIApplication application)
...
+    public void DidReceiveRegistrationToken (Firebase.CloudMessaging.Messaging messaging, string fcmToken)
+    {
+        // Monitor token generation: To be notified whenever the token is updated.
+        var token = fcmToken;
+        Log.LogDebug("OnNewToken: '{Token}'", token);
+        var appServices = IPlatformApplication.Current?.Services;
+        if (appServices == null)
+            return;
+
+        var rpcHub = appServices.GetService<RpcHub>();
```
Comment (crui3er) [inline src/dotnet/App.Maui/Platforms/iOS/AppDelegate.cs]: I think we can extract code block below and use both from Android and iOS platforms.

File: src/dotnet/App.Maui/Platforms/iOS/AppDelegate.cs [hunk 4]
```diff
@@ -42,6 +52,28 @@ public override void DidEnterBackground(UIApplication application)
...
+        var commander = appServices.GetService<ICommander>();
+        var sessionResolver = appServices.GetRequiredService<TrueSessionResolver>();
+        if (rpcHub != null && commander != null)
+            _ = Task.Run(async () => {
+                await rpcHub.WhenClientPeerConnected().ConfigureAwait(false);
```
Comment (crui3er) [inline src/dotnet/App.Maui/Platforms/iOS/AppDelegate.cs]: Add exception handling for background task and log an error if it occurs. May be, make sense to use our BackgrounTask.

**Review by crui3er — CHANGES_REQUESTED** [review]

**Pull request merged**